sql/core/src/main/scala/org/apache/spark/sql/DataFrameReader.scala

```diff
@@ -508,7 +508,9 @@ class DataFrameReader private[sql](sparkSession: SparkSession) extends Logging {
       sparkSession.sessionState.conf.sessionLocalTimeZone)
     val filteredLines: Dataset[String] =
       CSVUtils.filterCommentAndEmpty(csvDataset, parsedOptions)
-    val maybeFirstLine: Option[String] = filteredLines.take(1).headOption
+    val maybeFirstLine: Option[String] =
+      if (userSpecifiedSchema.isEmpty)
+	filteredLines.take(1).headOption else None
```

Comment:
This indent is off. I might just add the braces and newline for clarity.

Please see https://spark.apache.org/contributing.html ; this one probably needs a JIRA and other modifications.

But more to the point this changes the logic of `linesWithoutHeader`. This seems to skip handling the header too.

Reply:
Yea, this is used to skip first line as well even if users provide schema.

Reply:
Thanks guys for pointing that out. I pushed a new commit to the PR to address that.

To go into more detail. It seems that the first line scan accomplishes three goals:

- Schema inference
- Schema enforcement when schema is specified by the user. 
- Scanning the header string, so we know which lines to drop.

If the user specifies a schema and header is false, then none of the above implies. The `linesWithoutHeader` is just a computationally wasteful noop. When header is false, `CSVHeaderChecker` takes no action. And `CSVUtils` just passes through the unfiltered iterator. 

Therefore as long as schema is set, and header is false, then short-circuiting `maybeFirstLine` does not change any downstream behavior. 

And for that that, we get major efficiency gains. The entire executor (or entire cluster if dynamic allocation isn't enabled) is sitting idle, waiting for 1 core to run an entire job just to get 1 line.